### ISSUE TITLE
syntax: don't print trailing whitespace in comments

### ISSUE THREAD

**Issue by ebfe** [mvdan/sh#132]
```
$ cat x.sh | tr ' ' '_'
#_4_spaces:____
X="4_spaces:____
nextline"
$ shfmt x.sh | tr ' ' '_'
#_4_spaces:____
X="4_spaces:____
nextline"
```
Are there cases where these are not safe to remove?

**Comment by mvdan** [comment]
In the case of strings, I can think of plenty. You're changing the program - the string will be different.

I guess we could with comments, but it's sort of the same issue - what if you really wanted the space there? Although it couldn't possibly break a program, unless a program used itself as input.

Do any formatters like `gofmt` do any of this?

**Comment by ebfe** [comment]
> In the case of strings, I can think of plenty. You're changing the program - the string will be different.

Do you have an example? Trailing whitespace in quotes seems to be collapsed into a
single space:

```
$ cat x.sh |tr ' ' '_'
X="X________
X"
echo_$X
Y='Y_____
Y'
echo_$Y
Z="Z
Z"
echo_$Z
$ bash x.sh 
X X
Y Y
Z Z
```

> Do any formatters like `gofmt` do any of this?

gofmt removes trailing whitespace in comments.

**Comment by ebfe** [comment]
Scratch the quote part, yes arguments/quotes etc. Embarrasing ;)

**Comment by mvdan** [comment]
Yeah, we could do the comments. It just never occurred to me to be honest.

Thanks!

**Comment by mvdan** [comment]
Wondering if this should be hard-coded in the parser (i.e. trailing spaces in comments are ignored), hard-coded in the formatter, or optional only as part of `Simplify`.

**Comment by ebfe** [comment]
Formatter sounds like the right place, unless there are other reasons to not put it there?

**Comment by mvdan** [comment]
Do you mean the printer by this?

I think the parser would be a bad place, because otherwise the information is completely lost. Even if the printer removes spaces anywhere else, most of them can still be figured out by looking at node positions.

Also, this is how Go does it - `go/printer` uses `strings.TrimRightFunc(s, unicode.IsSpace)` on the comment text.

**Comment by mvdan** [comment]
Also, patch welcome of course. On holidays at the moment so might not get to this for a few days.

**Comment by mvdan** [comment]
Fixed in 604c59a91a90d3db88da267f59e0757824a9dd01 - we forgot the "fixes #n" in the commit message.
